src/Adapter/Swoole/EventLoop.php

```diff
@@ -0,0 +1,298 @@
+<?php
+
+namespace M6Web\Tornado\Adapter\Swoole;
+
+use Generator;
+use M6Web\Tornado\Adapter\Swoole\Internal\DummyPromise;
+use M6Web\Tornado\Deferred;
+use M6Web\Tornado\Promise;
+use Swoole\Coroutine;
+use RuntimeException;
+use Swoole\IDEHelper\StubGenerators\Swoole;
+use Swoole\Runtime;
+use Throwable;
+use function extension_loaded;
+
+class EventLoop implements \M6Web\Tornado\EventLoop
+{
+    private $cids;
+
+    public function __construct()
+    {
+        if (!extension_loaded('swoole')) {
+            throw new RuntimeException(
+                'EventLoop must running only with swoole extension.'
+            );
+        }
+
+        $this->cids = [];
+    }
+
+    private function shiftCoroutine(): mixed
+    {
+        if(count($this->cids) === 0) {
+            return null;
+        }
+
+        $cid = array_shift($this->cids);
+        if(Coroutine::exists($cid)) {
+            Coroutine::resume($cid);
+        } else {
+            $this->shiftCoroutine();
+        }
+
+        return $cid;
+    }
+
+    private function pushCoroutine(): mixed
+    {
+        $cid = Coroutine::getCid();
+        $this->cids[] = $cid;
+        Coroutine::yield();
+        return $cid;
+    }
+
+    private function createPromise(): DummyPromise
+    {
+        return new DummyPromise(function () {
+            $this->shiftCoroutine();
+        });
+    }
+
+    private function resolve($value)
```

Comment:
In which case `$value`  can be something else than a `DummyPromise`?

Reply:
in the case when `resolve` is called for `async` only, but not for `wait`.
note that resolve is private, just factoring parts.